File: sdk/ruby-holons/lib/holons/discovery_types.rb
# frozen_string_literal: true

module Holons
  LOCAL = 0
  PROXY = 1
  DELEGATED = 2

  SIBLINGS = 0x01
  CWD = 0x02
  SOURCE = 0x04
  BUILT = 0x08
  INSTALLED = 0x10
  CACHED = 0x20
  ALL = 0x3F

  NO_LIMIT = 0
  NO_TIMEOUT = 0

  IdentityInfo = Struct.new(
    :given_name,
    :family_name,
    :motto,
    :aliases,
    keyword_init: true
  )

  HolonInfo = Struct.new(
    :slug,
    :uuid,
    :identity,
    :lang,
    :runner,
    :status,
    :kind,
    :transport,
    :entrypoint,
    :architectures,
    :has_dist,
    :has_source,
    keyword_init: true
  )

  HolonRef = Struct.new(:url, :info, :error, keyword_init: true)
  DiscoverResult = Struct.new(:found, :error, keyword_init: true)
  ResolveResult = Struct.new(:ref, :error, keyword_init: true)
  ConnectResult = Struct.new(:channel, :uid, :origin, :error, keyword_init: true)

  begin
    require "grpc"
    @grpc_load_error = nil
  rescue LoadError => e
    @grpc_load_error = e
  end

  class << self
    def grpc_available?
      @grpc_load_error.nil?
    end

    def grpc_load_error
      @grpc_load_error
    end
  end
end

File: sdk/python-holons/holons/discovery_types.py
from __future__ import annotations

"""Uniform discovery API result types and constants."""

from dataclasses import dataclass, field

import grpc

# Scope flags.
LOCAL = 0
PROXY = 1
DELEGATED = 2

# Layer flags.
SIBLINGS = 0x01
CWD = 0x02
SOURCE = 0x04
BUILT = 0x08
INSTALLED = 0x10
CACHED = 0x20
ALL = 0x3F

# Clarity constants.
NO_LIMIT = 0
NO_TIMEOUT = 0


@dataclass
class IdentityInfo:
    given_name: str
    family_name: str
    motto: str = ""
    aliases: list[str] = field(default_factory=list)


@dataclass
class HolonInfo:
    slug: str
    uuid: str
    identity: IdentityInfo
    lang: str
    runner: str
    status: str
    kind: str
    transport: str
    entrypoint: str
    architectures: list[str]
    has_dist: bool
    has_source: bool


@dataclass
class HolonRef:
    url: str
    info: HolonInfo | None
    error: str | None


@dataclass
class DiscoverResult:
    found: list[HolonRef]
    error: str | None


@dataclass
class ResolveResult:
    ref: HolonRef | None
    error: str | None


@dataclass
class ConnectResult:
    channel: grpc.Channel | None
    uid: str
    origin: HolonRef | None
    error: str | None
